File: GrokkingDP/4/MinDelToPalindrome.cpp
#include <iostream>
#include <vector>

int MinDelToPalindrome(std::string str);

int main(void)
{
  std::cout << MinDelToPalindrome("abdcba") << ":"
            << MinDelToPalindrome("cddpd") << ":"
            << MinDelToPalindrome("pqr");
  return 0;
}

int MinDelToPalindrome(std::string str) {
  int n = str.length();
  std::vector<std::vector<int>> dp(n, std::vector<int>(n));
  for (int i = 0; i < n; ++i)
    dp[i][i] = 1;
  for (int l = n - 2; l >= 0; --l) {
    for (int r = l + 1; r < n; ++r) {
      int match = 0;
      if (str[l] == str[r]) match = dp[l + 1][r - 1] + 2;
      dp[l][r] = std::max(match, std::max(dp[l + 1][r], dp[l][r - 1]));
    }
  }
  return n - dp[0][n - 1];
}

File: GrokkingDP/4/min_del_to_palindrome.py
#  Problem Statement #
#  Given a string, find the minimum number of characters that we can remove to make it a palindrome.

#  Example 1:

#  Input: "abdbca"
#  Output: 1
#  Explanation: By removing "c", we get a palindrome "abdba".

#  Example 2:

#  Input: = "cddpd"
#  Output: 2
#  Explanation: Deleting "cp", we get a palindrome "ddd".

#  Example 3:

#  Input: = "pqr"
#  Output: 2
#  Explanation: We have to remove any two characters to get a palindrome, e.g. if we
#  remove "pq", we get palindrome "r".

def bu(s: str) -> int:
    """
    >>> bu('abdbca')
    1
    >>> bu('cddpd')
    2
    >>> bu('pqr')
    2
    """
    S, mx = len(s), 1
    dp = [[1 if r == c else 0 for c in range(S)] for r in range(S)]
    for i in range(S-1, -1, -1):
        for j in range(i+1, S):
            if s[i] == s[j]: dp[i][j] = dp[i+1][j-1] + 2
            else: dp[i][j] = max(dp[i+1][j], dp[i][j-1])
            mx = max(mx, dp[i][j])
    return S - mx

if __name__ == '__main__':
    import doctest
    doctest.testmod()
